lara-typescript/src/interactive-api-parent/modal-api-plugin.ts

```diff
@@ -0,0 +1,145 @@
+}
+const modalMap: Record<string, ModalState> = {};
+
+export function hasModal(uuid: string) {
+  return !!modalMap[uuid];
+}
+
+export function ModalApiPlugin(): IFrameSaverPlugin {
+  return {
+    listeners: {
+      showModal: (options: IShowModal) => {
+        const fnMap: Record<ModalType, (opts: IShowModal) => void> = {
+          alert: (opts: IShowAlert) => showAlert(opts),
+          lightbox: (opts: IShowLightbox) => showLightbox(opts),
+          dialog: (opts: IShowDialog) => showDialog(opts)
+        };
+        const showFn = fnMap[options.type];
+        showFn?.(options);
+      },
+      closeModal: (options: ICloseModal) => {
+        const fnMap: Record<ModalType, (opts: ICloseModal) => void> = {
+          alert: (opts: ICloseModal) => closeAlert(opts),
+          lightbox: (opts: ICloseModal) => closeLightbox(opts),
+          dialog: (opts: ICloseModal) => closeDialog(opts)
+        };
+        const type = modalMap[options.uuid]?.type;
+        const closeFn = type && fnMap[type];
+        closeFn?.(options);
+      }
+    }
+  };
+}
+
+function showAlert(options: IShowAlert) {
+  const { style, title: _title, text } = options;
+  let title: string;
+  let titlebarColor: string | undefined;
+  let message: string;
+  if (style === "correct") {
+    title = _title != null ? _title : "Correct";
+    titlebarColor = "#75a643";
+    message = text || "Yes! You are correct.";
+  }
+  else if (style === "incorrect") {
+    title = _title != null ? _title : "Incorrect";
+    titlebarColor = "#b45532";
+    message = text || "Sorry, that is incorrect.";
+  }
+  else {
+    title = _title || "";
+    message = text || "";
+  }
+
+  const $content = $(`<div class='check-answer'><p class='response'>${message}</p></div`);
+  modalMap[options.uuid] = { type: options.type, $content };
+  addPopup({
+    content: $content[0],
+    title,
+    titlebarColor,
+    modal: true,
+    onClose: () => delete modalMap[options.uuid]
+  });
+}
+
+function closeAlert(options: ICloseModal) {
+  const $content = modalMap[options.uuid]?.$content;
+  if ($content?.is(":ui-dialog")) {
+    $content.dialog("close");
+  }
+}
+
+function showLightbox(options: IShowLightbox) {
+  const getSizeOptions = () => {
+    const { isImage, size } = options;
+    // colorbox implementation detail
+    const kChromeSize = { width: 42, height: 70 };
+    const availableWidth = window.innerWidth - kChromeSize.width;
+    const availableHeight = window.innerHeight - kChromeSize.height;
+    const sizeOpts: { width?: number, height?: number, scalePhotos?: boolean } = {};
+    // images are auto-sized correctly without intervention
+    if (isImage) {
+      // prevent scaling up unless requested
+      const requiresUpscaling = (!!size?.width && size.width < availableWidth) &&
+                                (!!size?.height && size.height < availableHeight);
+      sizeOpts.scalePhotos = !requiresUpscaling || options.allowUpscale;
+      return sizeOpts;
+    }
```

Comment:
I didn't think we were going to support showing an image in this lightbox api. Even the image interactive will need to be an iframe in order to render it custom magnifying glass and caption text.

Reply:
The existing image functionality already supports rendering a magnifying glass and customizable caption text. If it's important to make the magnifying glass customizable as well we can add that capability (although I'm not sure why we'd want different interactives to show different magnifying glass icons).

The only current client of the lightbox API is primarily interested in showing an image. Toward that end, `<img>` tags have a lot of nice auto-sizing capabilities that `<iframe>` tags do not. I suspect there would be a fair amount of work involved to get the `<iframe>` implementation to behave as well as the `<img>` does now -- I started down that path and switched to supporting the image option when the sizing wasn't working. On top of that, the `<iframe>` implementation is more work on the client side as well. I'm not convinced that the additional effort is worth it at this juncture.

Reply:
It kind of seems like making the image into an interactive was kind of a waste then. If LARA is going to directly support showing image light boxes it might as well just render the image itself too. 

My hunch is that once we have the full iframe support required by the drawing tool, then we can remove this image specific api and just use that one. So I guess I'm OK with this approach for now, and I'll wait to see how that dialog api turns out.

Are the `<img>` autosizing features ones provided by Colorbox, or is it just that Colorbox is using these features? 
We have to implement these autosizing features for iframes anyhow. Currently that code is in iframe-sizing.js. I'm trying to get that refactored so at least some parts of it will be usable by React. 

I wonder how far you are going to get with Colorbox. I'd guess it would be better to switch to a React based lightbox/dialog. Since we will need that React lightbox/dialog in the activity player. So if we are putting in more than a little bit of time on Colorbox it is probably not a good thing.

Reply:
>My hunch is that once we have the full iframe support required by the drawing tool, then we can remove this image specific api and just use that one. So I guess I'm OK with this approach for now, and I'll wait to see how that dialog api turns out.

Maybe, but I wouldn't be surprised if full iframe support never reaches full parity with the image-specific API. If all you need to do is show an image, putting an `<iframe>` around an `<img>` tag just adds complexity. Separately, I took a quick look around for react lightbox libraries on npm -- there's a bunch of them that are tailored to work well with images but only one came up that supported non-image content (in addition to images). When we switch to a React-based implementation for the activity player I suspect the image functionality will be straightforward to achieve but the iframe support will be a substantial amount of work there as well.

>Are the `<img>` autosizing features ones provided by Colorbox, or is it just that Colorbox is using these features?

Colorbox simply uses an `<img>` tag when an image is requested and `<img>` tags resize their contents in convenient ways that `<iframe>`s don't.

>I wonder how far you are going to get with Colorbox. I'd guess it would be better to switch to a React based lightbox/dialog. Since we will need that React lightbox/dialog in the activity player. So if we are putting in more than a little bit of time on Colorbox it is probably not a good thing.

Which is exactly why I'm using the part of Colorbox that works without additional modification. I'm assuming that the jQuery-based implementation will only exist until the activity player gets far enough along, so there's no point in spending any more time on it than is required to achieve parity with the existing implementation.